strawberryfields/circuitspecs/xstrict.py

```diff
@@ -0,0 +1,183 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Circuit class specification for the X class of circuits."""
+
+import copy
+
+import numpy as np
+from thewalrus.quantum import Amat
+from thewalrus.symplectic import expand
+
+import strawberryfields as sf
+from strawberryfields.decompositions import takagi
+from strawberryfields.program_utils import CircuitError, Command, group_operations
+
+import strawberryfields.ops as ops
+
+from .circuit_specs import CircuitSpecs, Ranges
+from .gbs import GBSSpecs
+from .gaussian_unitary import GaussianUnitary
+
+
+class XstrictSpecs(CircuitSpecs):
+    """Circuit specifications for the X class of circuits.
+
+    An important property of this compilation routine is that it is done at the covariance matrix level.
+    This implies that one should not use it to compare the interferometers of a given circuit since they may
+    differ by permutations in the unitary and the squeezing parameters.
+    """
+
+    short_name = "Xstrict"
+    modes = None
+    remote = True
+    local = True
+    interactive = False
+    allowed_sq_ranges = Ranges([0], [1.0], variable_name="r")
+    sq_amplitude = 1.0
+
+    primitives = {
+        "S2gate",
+        "Sgate",
+        "MeasureFock",
+        "Rgate",
+        "BSgate",
+        "MZgate",
+        "Interferometer",
+    }
+
+    decompositions = {
+        "BipartiteGraphEmbed": {"mesh": "rectangular_symmetric", "drop_identity": False,},
+    }
+
+    def compile(self, seq, registers, strict=False):
+        # the number of modes in the provided program
+        n_modes = len(registers)
+
+        # Number of modes must be even
+        if n_modes % 2 != 0:
+            raise CircuitError("The X series only supports programs with an even number of modes.")
+        half_n_modes = n_modes // 2
+        # Call the GBS compiler to do basic measurement validation.
+        # The GBS compiler also merges multiple measurement commands
+        # into a single MeasureFock command at the end of the circuit.
+        seq = GBSSpecs().compile(seq, registers)
+
+        # ensure that all modes are measured
+        if len(seq[-1].reg) != n_modes:
+            raise CircuitError("All modes must be measured.")
+
+        # Check circuit begins with two-mode squeezers
+        # --------------------------------------------
+        A, B, C = group_operations(seq, lambda x: isinstance(x, ops.S2gate))
+        # If there are no two-mode squeezers add squeezers at the beginning with squeezing param equal to zero.
+        if B == []:
```

Comment:
```suggestion
        if not B:
```

Reply:
What happens if there are _some_ TMS gates, but not all? Should we be filling in all missing S2gates with `S2gate(0)`?

Reply:
Edit: nevermind, I see you are already doing this below :)